peps/pep-0735.rst

```diff
@@ -0,0 +1,613 @@
+a non-goal.
+
+Dependency Groups are not Hidden Extras
+---------------------------------------
+
+One could be forgiven for thinking that Dependency Groups are just extras which
+go unpublished.
+
+However, there are two major features which distinguish them from
+extras:
+
+* they support non-package projects
+
+* installation of a dependency group does not imply installation of a package's
+  dependencies (or the package itself)
+
+Object Specification Does Not Allow for "PEP 508 Decomposition"
+---------------------------------------------------------------
+
+Poetry and PDM both allow for their object formats for dependency data to be
+used to decompose a PEP 508 specification into parts. For example, a dependency
+on ``requests==2.21.0`` could be written under these tools in a form like
+``{name = "requests", version = "==2.21.0"}``.
+
+This spec does not allow for such a structure. There is only one way to write
+down a PEP 508 dependency, as a string. This has two positive effects:
+
+* There are fewer ways of writing identical data
+
+* It is not possible, by design, to combine a PEP 508 specifier with other
+  key-value pairs in a dependency specifier
+
+Specification
+=============
+
+This PEP defines a new section (table) in ``pyproject.toml`` files named
+``dependency-groups``. The ``dependency-groups`` table contains an arbitrary
+number of user-defined keys, each of which has, as its value, a list of
+requirements specifiers (defined below).  These keys must match the following
+regular expression: ``[a-z0-9][a-z0-9-]*[a-z0-9]``. Meaning that they must be
+all lower-case alphanumerics, with ``-`` allowed only in the middle, and at
+least two characters long. These requirements are chosen so that the
+normalization rules used for PyPI package names are unnecessary as the names
+are already normalized.
+
+Requirements specifiers will use a definition based on standardized
+`Dependency Specifiers <https://packaging.python.org/en/latest/specifications/dependency-specifiers/>`__
+introduced in :pep:`508`.
+This PEP also proposes an object format to define a dependency, called a "PEP
+735 Dependency" (defined below). The elements in ``[dependency-groups]``
+lists must either be strings, in which case they must be valid Dependency
+Specifiers or else PEP 735 Dependencies.
+
+PEP 735 Dependencies
+--------------------
+
+PEP 735 Dependencies are objects which define a set of dependencies. They do
+not necessarily refer to a single package, as will become clear from the
+definitions below.
+
+There are two forms of PEP 735 Dependencies: "Dependency Group Includes" and
+"Path Dependencies".
+
+Dependency Group Include
+''''''''''''''''''''''''
+
+A Dependency Group Include includes the dependencies of another Dependency
+Group in the current Dependency Group.
+
+An include is defined as an object with exactly one key, ``"include"``, whose
+value is a string, the name of another Dependency Group.
```

Comment:
Could this possibly be made a list? Just asking, maybe I've missed a discussion.

Reply:
There hasn't been discussion of this yet.
I think a list is pretty near equivalent, but I was concerned that it's adding work for implementers without adding much value for users.

I'd like to leave it as a string for the moment, but I'll add this to open issues so that it doesn't slip from discussion.

Reply:
Let's keep non-editorial discussion separate from this PR, and in the discuss.python.org thread for this PR please.